### ISSUE TITLE
Large divergence between eager and XLA for sin/cos on large inputs

### ISSUE THREAD

**Issue by beanduan22** [tensorflow/tensorflow#115507]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.11

### Custom code

Yes

### Python version

3.10

### Current behavior?

For large-magnitude inputs, TensorFlow produces significantly different results between eager execution and XLA-compiled execution when using transcendental functions such as tf.sin and tf.cos.

In our test, the same model produces outputs differing by up to 1e-1 between eager and XLA, which is far beyond typical floating-point tolerance.

This discrepancy appears to originate from differences in argument reduction or approximation strategies for sin/cos under large inputs, and is further amplified by downstream operations (e.g., BatchNorm and reductions).

Expected behavior:
Eager and XLA execution should produce numerically consistent results within reasonable floating-point tolerance. Differences of this magnitude may affect model outputs and downstream decisions.

### Standalone code to reproduce the issue

```shell
import os; os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import tensorflow as tf
import numpy as np

tf.random.set_seed(42)

class SinCosBNModel(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.dense1 = tf.keras.layers.Dense(8)
        self.bn = tf.keras.layers.BatchNormalization()

    def call(self, x, training=False):
        x = self.dense1(x)
        x = tf.sin(x)
        x = self.bn(x, training=training)
        return tf.reduce_sum(tf.cos(x))

model = SinCosBNModel()
tf.keras.utils.set_random_seed(42)
x_init = tf.constant(np.random.RandomState(42).randn(4, 8).astype(np.float32))
_ = model(x_init, training=True)

@tf.function(jit_compile=True)
def xla_forward(x):
    return model(x, training=False)

def eager_forward(x):
    return model(x, training=False)

# Search for an input that triggers the divergence
tol = 1e-6
x_large = None
max_diff = 0.0
eager_out = xla_out = None

for seed in range(50):
    x_base = tf.constant(np.random.RandomState(seed).randn(4, 8).astype(np.float32))
    for scale in [100.0, 50.0, 200.0, 1000.0]:
        candidate = x_base * scale
        e = eager_forward(candidate)
        xl = xla_forward(candidate)
        d = float(tf.abs(e - xl))
        if d > max_diff:
            max_diff = d
            x_large = candidate
            eager_out = e
            xla_out = xl
        if d > tol:
            break
    if max_diff > tol:
        break

print(f"Eager: {eager_out.numpy():.8f}")
print(f"XLA:   {xla_out.numpy():.8f}")
print(f"Max diff = {max_diff:.3e}, tol = {tol:.0e}")

if max_diff > tol:
    print("PASS")
else:
    print("NOT REPRODUCED")
```

### Relevant log output

```shell
Eager: 26.77629662
XLA:   26.92271042
Max diff = 1.464e-01, tol = 1e-06
```

**Comment by Venkat6871** [comment]
Hi @beanduan22,

Apologies for the delay, and thanks for the detailed report and reproducible example.
Since this behavior appears with `jit_compile=True`, this looks more like an XLA-related numerical difference than a confirmed correctness bug at this point. XLA does not guarantee identical results to eager execution for floating-point computations, and differences can become more noticeable with large-magnitude inputs and when combined with downstream ops.

To better isolate the source, could you try a minimal example that compares `tf.sin` / `tf.cos` directly on the same input, without Dense, BatchNormalization, or reductions? That would help confirm whether the divergence is coming from the transcendental ops themselves or from accumulated numerical effects in the larger graph.

I also tried a minimal version of this setup on a more recent TensorFlow version and did not observe a significant difference. Sharing a small [gist](https://colab.sandbox.google.com/gist/Venkat6871/72eab1764bafb44755f8a8508a0a06e9/115507_tf_2-21-nightly-v.ipynb) for reference in case it helps.

Also, if possible, please try with a more recent TensorFlow version, since this report is based on 2.11.

Thank you!

**Comment by beanduan22** [comment]
Thanks for the suggestion — I agree that differences between eager and XLA can arise from floating-point behavior.

However, in this case the magnitude of the discrepancy (~1e-1) under finite inputs seems larger than typical numerical variation. Since sin/cos involve argument reduction for large inputs, even small differences in reduction or approximation can lead to noticeable output drift.

The concern here is that these differences are further amplified by downstream operations (e.g., normalization and reductions), which may impact model outputs in practice.

Would it make sense to consider whether large-magnitude inputs for transcendental ops might require more stable handling.

Thanks again for taking a look!
